ui/lib/components/autocomplete/autocomplete.dart

```diff
@@ -0,0 +1,123 @@
+import 'dart:html';
+
+class SuggestionItem {
+  String _value;
+  String _searchString;
+  HtmlElement _display;
+
+  String get value => _value;
+  String get searchString => _searchString;
+
+  SuggestionItem(this._value, this._searchString, this._display);
+}
+
+class AutocompleteList {
+  DivElement renderElement;
+  DivElement _suggestionsList;
+  DivElement _emptyPlaceholder;
+
+  int _listFocusIndex;
+
+  List<SuggestionItem> _allSuggestions;
+  List<SuggestionItem> get activeSuggestions {
+    if (_inputText == "") return _allSuggestions;
+    return _allSuggestions.where((suggestionItem) {
+      return suggestionItem.searchString.toLowerCase().contains(_inputText.toLowerCase());
+    }).toList();
+  }
+
+  String _inputText = "";
+  void set inputText(String input) {
+    _inputText = input;
+    if (_inputText == "") {
+      _listFocusIndex = null;
+    } else if (_listFocusIndex != null) {
+      _listFocusIndex = 0;
+    }
+    _updateSuggestionListRender();
+  }
+
+  void Function(SuggestionItem) onSelect = (_) {};
+  void Function() onRequestClose = () {};
+  void Function() onFocus = () {};
+  void Function() onBlur = () {};
+
+  AutocompleteList(this._allSuggestions, this._inputText, {DivElement emptyPlaceholder}) {
+    _emptyPlaceholder = emptyPlaceholder ?? DivElement();
+
+    renderElement = DivElement()..classes.add("autocomplete__wrapper");
+    _suggestionsList = DivElement()..classes.add("autocomplete__suggestions");
+    renderElement.append(_suggestionsList);
+    _updateSuggestionListRender();
+
+    document.onKeyUp.listen(_handleKeyboardInteraction);
+  }
+
+  void _handleKeyboardInteraction(KeyboardEvent event) {
+    if (activeSuggestions.isEmpty) return;
+    switch (event.key) {
+      case "ArrowUp":
+        event.preventDefault(); // to prevent page scrolling
+        _listFocusIndex = _listFocusIndex == null ? activeSuggestions.length - 1 : _listFocusIndex - 1;
+        break;
+      case "ArrowDown":
+        event.preventDefault(); // to prevent page scrolling
+        _listFocusIndex = _listFocusIndex == null ? 0 : _listFocusIndex + 1;
+        break;
+      case "Escape":
+        _listFocusIndex = null;
+        onRequestClose();
+        break;
+      case "Enter":
+        if (_listFocusIndex != null && _listFocusIndex >= 0 && _listFocusIndex < activeSuggestions.length) {
+          _onChoose(activeSuggestions[_listFocusIndex]);
+        }
+        break;
+      default:
+        return;
```

Comment:
wouldn't this `return` mean that if the users presses something else other than arrow/enter/escape, then that key won't get processed and, e.g. the suggested list won't get updated? It feels like almost we should `return` on arrow/enter/escape, and `break` by default. Am I missing something?

Reply:
Ah so the autocomplete doesn't carry any input fields. All it does is takes an input string and when asked to show the popup, it displays the appropriate options. And this popup listens to the arrow keys and enter key to choose the options. This way we can attach the autocomplete component to any input element (text area, input type=text, contenteditable div / span, select). Please let me know if that clarifies it.

Reply:
ah, yes, that makes sense - this is just the list, not the input element!  thanks for the explanation, LGTM!